python/cudf/cudf/_lib/spillable_buffer.pyx

```diff
@@ -0,0 +1,448 @@
+
+        Parameters
+        ----------
+        target : str
+            The target of the spilling.
+        """
+
+        assert self._view_desc is None
+        with self._lock:
+            ptr_type = self._ptr_desc["type"]
+            if ptr_type == target:
+                return
+
+            if not self.spillable:
+                raise ValueError(
+                    f"Cannot in-place move an unspillable buffer: {self}"
+                )
+
+            if (ptr_type, target) == ("gpu", "cpu"):
+                host_mem = memoryview(bytearray(self.size))
+                rmm._lib.device_buffer.copy_ptr_to_host(self._ptr, host_mem)
+                self._ptr_desc["memoryview"] = host_mem
+                self._ptr = 0
+                self._owner = None
+            elif (ptr_type, target) == ("cpu", "gpu"):
+                # Notice, this operation is prone to deadlock because the RMM
+                # allocation might trigger spilling-on-demand which in turn
+                # trigger a new call to this buffer's `__spill__()`.
+                # Therefore, it is important that spilling-on-demand doesn't
+                # tries to unspill an already locked buffer!
+                dev_mem = rmm.DeviceBuffer.to_device(
+                    self._ptr_desc.pop("memoryview")
+                )
+                self._ptr = dev_mem.ptr
+                self._size = dev_mem.size
+                self._owner = dev_mem
+            else:
+                # TODO: support moving to disk
+                raise ValueError(f"Unknown target: {target}")
+            self._ptr_desc["type"] = target
+
+    @property
+    def ptr(self) -> int:
+        """Access the memory directly
+
+        Notice, this will mark the buffer as "exposed" and make
+        it unspillable permanently.
+
+        Consider using `ptr_raw() or `.restricted_ptr()` instead.
+        """
+        if self._view_desc:
+            return self._view_desc["base"].ptr + self._view_desc["offset"]
+        self._manager.spill_to_device_limit()
+        with self._lock:
+            if not self._exposed:
+                self._manager.log_expose(self)
+            self.__spill__(target="gpu")
+            self._exposed = True
+            self._last_accessed = time.monotonic()
+            return self._ptr
+
+    cdef void* ptr_raw(self, SpillLock spill_lock) except *:
+        if spill_lock is None:
+            return <void*><uintptr_t> self.ptr
+
+        # Get base buffer
+        cdef SpillableBuffer base
+        cdef size_t offset
+        if self._view_desc is None:
+            base = self
+            offset = 0
+        else:
+            base = self._view_desc["base"]
+            offset = self._view_desc["offset"]
+
+        with base._lock:
+            base.__spill__(target="gpu")
+            base._last_accessed = time.monotonic()
+            spill_lock.add(base._expose_counter)
+            return <void*><uintptr_t> (base._ptr+offset)
```

Comment:
I guess not quite exactly, because `ptr_restricted` creates an internal `SpillLock`. It should still be possible to make this work by just having a default value for the lock parameter though I think.

Reply:
Merged `ptr_raw()` and `ptr_restricted()` into a new function `get_ptr()`: https://github.com/rapidsai/cudf/pull/11553/commits/35bf6f8eaded9e27a23ba5ef89d6145425ef4dce